add Mesh::isync_array to perform asynchronous update of arrays

tristan0x opened the pull request:
Now that I am creating the PR, I wonder if it wouldn't be better to move the definition of class `Omega_h::IRequest` in a cpp file...

ibaned commented:
@tristan0x thanks for this work! There are people at Sandia that want this too.

Moving `IRequest` to a `.cpp` is fine, I won't require it before acceptance.

I do have some questions though:

1. how do you ensure the lifetime of the send buffer persists until the communication is over? I see that `IRequest` contains either a copy or a reference to the receive buffer... I guess I think `IRequest` should keep a copy of both the send and receive buffers (since they have `std::shared_ptr` semantics) to ensure that they are not deallocated before the communication phase is over.

2. shouldn't there be some post-processing of the receive buffer? i.e. permutation from sorted-by-rank to sorted-by-destination index (see the code in `Dist::exch` after `alltoallv` call)? Should the `IRequest` do this?

3. I think `IRequest` looks at lot like `std::future`... in fact it is better than `std::future`, because the method `completed` seems like a better asynchronous interface than `wait`. Anyways, I'm not sure if I want to name it `Omega_h::Future` or anything, but just thinking out loud...

tristan0x commented on src/Omega_h_comm.cpp:
```diff
@@ -488,20 +534,22 @@ Read<T> Comm::alltoallv(Read<T> sendbuf_dev, Read<LO> sdispls_dev,
+  auto const& reqs = Neighbor_ialltoallv(host_srcs_, host_dsts_, width,
       nonnull(sendbuf.data()), nonnull(sdispls.data()),
       MpiTraits<T>::datatype(), nonnull(recvbuf.data()),
       nonnull(rdispls.data()), MpiTraits<T>::datatype(), impl_));
```
Thank you @bgranzow 
I didn't have the chance to test CUDA compilation. I submitting a fix based on a dirty (#define OMEGA_H_USE_CUDA 1 in the .cpp file...). Proper tests will be done before merging this PR.

tristan0x commented on src/Omega_h_comm.cpp:
```diff
@@ -488,20 +534,22 @@ Read<T> Comm::alltoallv(Read<T> sendbuf_dev, Read<LO> sdispls_dev,
   HostRead<T> sendbuf(sendbuf_dev);
   OMEGA_H_CHECK(recvbuf.size() == rdispls.last() * width);
-  CALL(Neighbor_alltoallv(host_srcs_, host_dsts_, width,
+  auto const& reqs = Neighbor_ialltoallv(host_srcs_, host_dsts_, width,
```
thank you. this has been addressed in 2389576

tristan0x commented on src/Omega_h_comm.cpp:
```diff
@@ -471,6 +471,52 @@ void self_send_part2(Read<T> self_data, LO self_src, Read<T>* p_recvbuf,
+  (void)rdispls_dev;
+  (void)width;
+  IRequest<T>::requests_type reqs;
+  auto future = [&]() { return sendbuf_dev; };
```
It tells to capture all symbols of the scope by reference. `=` captures everything by copy.

tristan0x commented:
> 2\. My preference would be that you not modify any of the internals of `Comm::alltoallv` so that the old MPI behavior is completely retained. That appeals to me because there aren't that many MPI tests in the current test harness, and I could envision some downstream effects. Is there a reason that function was modified in the first place?

Indeed I changed `Comm::alltoallv` but the execution flow remains the same.
I removed the MPIWait from `Neighbor_ialltoallv` static function to reuse it in `ialltoallv`. It is now the responsibility of the caller to do the MPIWait

bgranzow commented:
> Indeed I changed Comm::alltoallv but the execution flow remains the same.
I removed the MPIWait from Neighbor_ialltoallv static function to reuse it in ialltoallv. It is now the responsibility of the caller to do the MPIWait

I didn't realize you took out `Neighbor_alltoallv`. My preference would be that you reinstate the function `Neighbor_alltoallv` and revert `Comm::alltoallv` back to calling it (even if `Neighbor_alltoallv` and `Neighbor_ialltoallv` are essentially duplicated code). Maybe @ibaned could weigh in on that.

ibaned commented:
I understand the concerns about correctness and I have them too but its better to set up some MPI testing at Sandia (*cough* with FLEXO funding) and ditch old implementations than to go down the Albany path of keeping all the old implementations around. It seems like the MPI code is headed for some inevitable modernization after this, and we'll need testing for every step of that transformation.

As for type traits, `Tag` is what does type erasure of arrays on a `Mesh`, but it is fairly closely tied to the `Mesh`. It may be possible to do similar type erasure where `IRequestTyped<T>` derives from `IRequestBase` (names for reference only).

bgranzow commented:
Well, I've been working on doing exactly that in FLEXO (setting up some MPI testing) and I could try to cook up some in Omega_h as well.

It was my intent to propose only keeping deprecated code until the above testing has actually been implemented and can be relied upon, then totally removing it.

That being said, @tristan0x's latest commits compile and all the tests pass with CUDA as a backend with CUDA_AWARE_MPI=OFF on the dgx machine, so that's pretty awesome.

bgranzow commented:
@  2389576 

with Clang 7.0.0

```
clang version 7.0.0 (tags/RELEASE_700/final)
Target: x86_64-apple-darwin16.7.0
Thread model: posix
InstalledDir: /Users/bngranz/utils/install/clang-7.0.0/bin
```

there are compilation errors shown below. The first few could probably be fixed by static_cast'ing the ints appropriately, I'm less clear about the destrutor errors.

```
/Users/bngranz/jughead/src/omega_h/src/Omega_h_comm.cpp:331:55: error: implicit conversion changes signedness: 'int' to 'std::__1::vector<int, std::__1::allocator<int> >::size_type' (aka 'unsigned long') [-Werror,-Wsign-conversion]
  IRequest<I32>::requests_type sendrecvreqs(outdegree + indegree);
                               ~~~~~~~~~~~~ ~~~~~~~~~~^~~~~~~~~~
/Users/bngranz/jughead/src/omega_h/src/Omega_h_comm.cpp:551:25: error: implicit conversion loses integer precision: 'std::__1::vector<int, std::__1::allocator<int> >::size_type' (aka 'unsigned long') to 'int' [-Werror,-Wshorten-64-to-32]
  CALL(MPI_Waitall(reqs.size(), reqs.data(), MPI_STATUSES_IGNORE));
       ~~~~~~~~~~~ ~~~~~^~~~~~
/Users/bngranz/jughead/src/omega_h/src/Omega_h_comm.cpp:19:47: note: expanded from macro 'CALL'
#define CALL(f) OMEGA_H_CHECK(MPI_SUCCESS == (f))
                                              ^
/Users/bngranz/jughead/src/omega_h/src/Omega_h_fail.hpp:35:5: note: expanded from macro 'OMEGA_H_CHECK'
  ((cond) ? ((void)0)                                                          \
    ^~~~
/Users/bngranz/jughead/src/omega_h/src/Omega_h_comm.cpp:582:1: note: in instantiation of function template specialization 'Omega_h::Comm::alltoallv<signed char>' requested here
INST(I8)
^
/Users/bngranz/jughead/src/omega_h/src/Omega_h_comm.cpp:577:26: note: expanded from macro 'INST'
  template Read<T> Comm::alltoallv(                                            \
                         ^
/Users/bngranz/jughead/src/omega_h/src/Omega_h_comm.cpp:551:25: error: implicit conversion loses integer precision: 'std::__1::vector<int, std::__1::allocator<int> >::size_type' (aka 'unsigned long') to 'int' [-Werror,-Wshorten-64-to-32]
  CALL(MPI_Waitall(reqs.size(), reqs.data(), MPI_STATUSES_IGNORE));
       ~~~~~~~~~~~ ~~~~~^~~~~~
/Users/bngranz/jughead/src/omega_h/src/Omega_h_comm.cpp:19:47: note: expanded from macro 'CALL'
#define CALL(f) OMEGA_H_CHECK(MPI_SUCCESS == (f))
                                              ^
/Users/bngranz/jughead/src/omega_h/src/Omega_h_fail.hpp:35:5: note: expanded from macro 'OMEGA_H_CHECK'
  ((cond) ? ((void)0)                                                          \
    ^~~~
/Users/bngranz/jughead/src/omega_h/src/Omega_h_comm.cpp:583:1: note: in instantiation of function template specialization 'Omega_h::Comm::alltoallv<int>' requested here
INST(I32)
^
/Users/bngranz/jughead/src/omega_h/src/Omega_h_comm.cpp:577:26: note: expanded from macro 'INST'
  template Read<T> Comm::alltoallv(                                            \
                         ^
/Users/bngranz/jughead/src/omega_h/src/Omega_h_comm.cpp:551:25: error: implicit conversion loses integer precision: 'std::__1::vector<int, std::__1::allocator<int> >::size_type' (aka 'unsigned long') to 'int' [-Werror,-Wshorten-64-to-32]
  CALL(MPI_Waitall(reqs.size(), reqs.data(), MPI_STATUSES_IGNORE));
       ~~~~~~~~~~~ ~~~~~^~~~~~
/Users/bngranz/jughead/src/omega_h/src/Omega_h_comm.cpp:19:47: note: expanded from macro 'CALL'
#define CALL(f) OMEGA_H_CHECK(MPI_SUCCESS == (f))
                                              ^
/Users/bngranz/jughead/src/omega_h/src/Omega_h_fail.hpp:35:5: note: expanded from macro 'OMEGA_H_CHECK'
  ((cond) ? ((void)0)                                                          \
    ^~~~
/Users/bngranz/jughead/src/omega_h/src/Omega_h_comm.cpp:584:1: note: in instantiation of function template specialization 'Omega_h::Comm::alltoallv<long long>' requested here
INST(I64)
^
/Users/bngranz/jughead/src/omega_h/src/Omega_h_comm.cpp:577:26: note: expanded from macro 'INST'
  template Read<T> Comm::alltoallv(                                            \
                         ^
/Users/bngranz/jughead/src/omega_h/src/Omega_h_comm.cpp:551:25: error: implicit conversion loses integer precision: 'std::__1::vector<int, std::__1::allocator<int> >::size_type' (aka 'unsigned long') to 'int' [-Werror,-Wshorten-64-to-32]
  CALL(MPI_Waitall(reqs.size(), reqs.data(), MPI_STATUSES_IGNORE));
       ~~~~~~~~~~~ ~~~~~^~~~~~
/Users/bngranz/jughead/src/omega_h/src/Omega_h_comm.cpp:19:47: note: expanded from macro 'CALL'
#define CALL(f) OMEGA_H_CHECK(MPI_SUCCESS == (f))
                                              ^
/Users/bngranz/jughead/src/omega_h/src/Omega_h_fail.hpp:35:5: note: expanded from macro 'OMEGA_H_CHECK'
  ((cond) ? ((void)0)                                                          \
    ^~~~
/Users/bngranz/jughead/src/omega_h/src/Omega_h_comm.cpp:585:1: note: in instantiation of function template specialization 'Omega_h::Comm::alltoallv<double>' requested here
INST(Real)
^
/Users/bngranz/jughead/src/omega_h/src/Omega_h_comm.cpp:577:26: note: expanded from macro 'INST'
  template Read<T> Comm::alltoallv(                                            \
                         ^
In file included from /Users/bngranz/jughead/src/omega_h/src/Omega_h_comm.cpp:1:
In file included from /Users/bngranz/jughead/src/omega_h/src/Omega_h_comm.hpp:11:
/Users/bngranz/jughead/src/omega_h/src/Omega_h_irequest.hpp:180:56: error: declaration requires an exit-time destructor [-Werror,-Wexit-time-destructors]
typename IRequest<T>::callback_type const IRequest<T>::identity_cb =
                                                       ^
/Users/bngranz/jughead/src/omega_h/src/Omega_h_irequest.hpp:52:38: note: in instantiation of static data member 'Omega_h::IRequest<signed char>::identity_cb' requested here
      const callback_type callback = identity_cb)
                                     ^
/Users/bngranz/jughead/src/omega_h/src/Omega_h_irequest.hpp:180:56: error: declaration requires a global destructor [-Werror,-Wglobal-constructors]
typename IRequest<T>::callback_type const IRequest<T>::identity_cb =
                                                       ^
/Users/bngranz/jughead/src/omega_h/src/Omega_h_irequest.hpp:184:61: error: declaration requires an exit-time destructor [-Werror,-Wexit-time-destructors]
typename IRequest<T>::host_callback_type const IRequest<T>::host_identity_cb =
                                                            ^
/Users/bngranz/jughead/src/omega_h/src/Omega_h_irequest.hpp:58:24: note: in instantiation of static data member 'Omega_h::IRequest<signed char>::host_identity_cb' requested here
        host_callback_(host_identity_cb),
                       ^
/Users/bngranz/jughead/src/omega_h/src/Omega_h_comm.cpp:509:10: note: in instantiation of member function 'Omega_h::IRequest<signed char>::IRequest' requested here
  return {sendbuf_dev, recvbuf_dev_w, std::move(reqs)};
         ^
In file included from /Users/bngranz/jughead/src/omega_h/src/Omega_h_comm.cpp:1:
In file included from /Users/bngranz/jughead/src/omega_h/src/Omega_h_comm.hpp:11:
/Users/bngranz/jughead/src/omega_h/src/Omega_h_irequest.hpp:184:61: error: declaration requires a global destructor [-Werror,-Wglobal-constructors]
typename IRequest<T>::host_callback_type const IRequest<T>::host_identity_cb =
                                                            ^
/Users/bngranz/jughead/src/omega_h/src/Omega_h_irequest.hpp:180:56: error: declaration requires an exit-time destructor [-Werror,-Wexit-time-destructors]
typename IRequest<T>::callback_type const IRequest<T>::identity_cb =
                                                       ^
/Users/bngranz/jughead/src/omega_h/src/Omega_h_irequest.hpp:52:38: note: in instantiation of static data member 'Omega_h::IRequest<int>::identity_cb' requested here
      const callback_type callback = identity_cb)
                                     ^
/Users/bngranz/jughead/src/omega_h/src/Omega_h_irequest.hpp:180:56: error: declaration requires a global destructor [-Werror,-Wglobal-constructors]
typename IRequest<T>::callback_type const IRequest<T>::identity_cb =
                                                       ^
/Users/bngranz/jughead/src/omega_h/src/Omega_h_irequest.hpp:184:61: error: declaration requires an exit-time destructor [-Werror,-Wexit-time-destructors]
typename IRequest<T>::host_callback_type const IRequest<T>::host_identity_cb =
                                                            ^
/Users/bngranz/jughead/src/omega_h/src/Omega_h_irequest.hpp:58:24: note: in instantiation of static data member 'Omega_h::IRequest<int>::host_identity_cb' requested here
        host_callback_(host_identity_cb),
                       ^
/Users/bngranz/jughead/src/omega_h/src/Omega_h_comm.cpp:509:10: note: in instantiation of member function 'Omega_h::IRequest<int>::IRequest' requested here
  return {sendbuf_dev, recvbuf_dev_w, std::move(reqs)};
         ^
In file included from /Users/bngranz/jughead/src/omega_h/src/Omega_h_comm.cpp:1:
In file included from /Users/bngranz/jughead/src/omega_h/src/Omega_h_comm.hpp:11:
/Users/bngranz/jughead/src/omega_h/src/Omega_h_irequest.hpp:184:61: error: declaration requires a global destructor [-Werror,-Wglobal-constructors]
typename IRequest<T>::host_callback_type const IRequest<T>::host_identity_cb =
                                                            ^
/Users/bngranz/jughead/src/omega_h/src/Omega_h_irequest.hpp:180:56: error: declaration requires an exit-time destructor [-Werror,-Wexit-time-destructors]
typename IRequest<T>::callback_type const IRequest<T>::identity_cb =
                                                       ^
/Users/bngranz/jughead/src/omega_h/src/Omega_h_irequest.hpp:52:38: note: in instantiation of static data member 'Omega_h::IRequest<long long>::identity_cb' requested here
      const callback_type callback = identity_cb)
                                     ^
/Users/bngranz/jughead/src/omega_h/src/Omega_h_irequest.hpp:180:56: error: declaration requires a global destructor [-Werror,-Wglobal-constructors]
typename IRequest<T>::callback_type const IRequest<T>::identity_cb =
                                                       ^
/Users/bngranz/jughead/src/omega_h/src/Omega_h_irequest.hpp:184:61: error: declaration requires an exit-time destructor [-Werror,-Wexit-time-destructors]
typename IRequest<T>::host_callback_type const IRequest<T>::host_identity_cb =
                                                            ^
/Users/bngranz/jughead/src/omega_h/src/Omega_h_irequest.hpp:58:24: note: in instantiation of static data member 'Omega_h::IRequest<long long>::host_identity_cb' requested here
        host_callback_(host_identity_cb),
                       ^
/Users/bngranz/jughead/src/omega_h/src/Omega_h_comm.cpp:509:10: note: in instantiation of member function 'Omega_h::IRequest<long long>::IRequest' requested here
  return {sendbuf_dev, recvbuf_dev_w, std::move(reqs)};
         ^
In file included from /Users/bngranz/jughead/src/omega_h/src/Omega_h_comm.cpp:1:
In file included from /Users/bngranz/jughead/src/omega_h/src/Omega_h_comm.hpp:11:
/Users/bngranz/jughead/src/omega_h/src/Omega_h_irequest.hpp:184:61: error: declaration requires a global destructor [-Werror,-Wglobal-constructors]
typename IRequest<T>::host_callback_type const IRequest<T>::host_identity_cb =
                                                            ^
/Users/bngranz/jughead/src/omega_h/src/Omega_h_irequest.hpp:180:56: error: declaration requires an exit-time destructor [-Werror,-Wexit-time-destructors]
typename IRequest<T>::callback_type const IRequest<T>::identity_cb =
                                                       ^
/Users/bngranz/jughead/src/omega_h/src/Omega_h_irequest.hpp:52:38: note: in instantiation of static data member 'Omega_h::IRequest<double>::identity_cb' requested here
      const callback_type callback = identity_cb)
                                     ^
/Users/bngranz/jughead/src/omega_h/src/Omega_h_irequest.hpp:180:56: error: declaration requires a global destructor [-Werror,-Wglobal-constructors]
typename IRequest<T>::callback_type const IRequest<T>::identity_cb =
```

tristan0x commented:
> That being said, @tristan0x's latest commits compile and all the tests pass with CUDA as a backend with CUDA_AWARE_MPI=OFF on the dgx machine, so that's pretty awesome.

I have Omega_h CMake issues on dgx.
Could you please share the CMake command you are using (or share your CMakeCache.txt) on dgx machine so that I can at least compile omega_h and run the tests?

tristan0x commented:
I have been able to build with both MPI and CUDA, thanks  @bgranzow !
Unless I am mistaken, I addressed all your remarks.

bgranzow commented:
Awesome, thanks @tristan0x!

I was able to build on my mac with clang 7.0.0, and on a GPU test-bed with CUDA aware MPI now turned on and all Omega_h tests passed and all downstream FLEXO MPI tests that I've been developing passed as well.

@ibaned, I'll merge this into develop if you're OK with that.

tristan0x commented:
Thank you @bgranzow 
I cleaned up the branch. I kept the first commit and squashed the others.
Best

ibaned commented:
Fine with me! Thanks again @tristan0x !
